### PR TITLE
Correct CA Young Child Tax Credit (YCTC) logic

### LINKED ISSUE #1870: CA YCTC expanded in 2022
This CA [webpage](https://www.ftb.ca.gov/forms/whats-new.html) contains the following notice:

Screenshot 2023-03-07 at 12 57 52 PM

The current code requires that a tax unit receive a positive CalEITC benefit, which was true before the 2022 tax year.

In addition to this needed enhancement, the `ca_yctc` formula uses the wrong earnings variable.
This bug needs fixing.

### PR DESCRIPTION
**Pull request by martinholmer**
Fixes #1870.

### CODE REVIEW

File: policyengine_us/tests/policy/baseline/gov/states/ca/tax/income/credits/young_child/ca_yctc.yaml [hunk 1]
```diff
@@ -112,11 +113,52 @@
...
+        members: [parent, child]
+    households:
+      household:
+        members: [parent, child]
+        state_code: CA
+  output:
+    # ca_eitc: 0  # because of zero earnings
+    ca_yctc: 0  # because does not get CalEITC
+
+- name: Zero earnings with losses, but still eligible in 2022.
```
Comment (MaxGhenis) [inline policyengine_us/tests/policy/baseline/gov/states/ca/tax/income/credits/young_child/ca_yctc.yaml:145]: Could you add a case that fails the net losses test?

File: policyengine_us/variables/gov/states/ca/tax/income/credits/earned_income/ca_eitc.py [hunk 2]
```diff
@@ -89,6 +89,17 @@ def modify_parameters(parameters: ParameterNode) -> ParameterNode:
         return simulation.branches.get("ca_eitc")
 
 
+class ca_eitc_eligible(Variable):
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/ca/tax/income/credits/earned_income/ca_eitc.py:92]: Could you move this to `ca_eitc_eligible.py` and import the branch from here?

Alternatively we could start making branch-making functions their own files, like `_get_ca_eitc_branch.py`. @nikhilwoodruff wdyt?

File: policyengine_us/variables/gov/states/ca/tax/income/credits/earned_income/ca_eitc.py [hunk 3]
```diff
@@ -89,6 +89,17 @@ def modify_parameters(parameters: ParameterNode) -> ParameterNode:
...
+    value_type = bool
+    entity = TaxUnit
+    label = "Eligible for CalEITC"
+    definition_period = YEAR
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/ca/tax/income/credits/earned_income/ca_eitc.py:96]:
```suggestion
    definition_period = YEAR
    defined_for = StateCode.CA
```

File: policyengine_us/variables/gov/states/ca/tax/income/credits/earned_income/ca_eitc.py [hunk 4]
```diff
@@ -89,6 +89,17 @@ def modify_parameters(parameters: ParameterNode) -> ParameterNode:
...
+
+    def formula(tax_unit, period, parameters):
+        branch = get_ca_eitc_branch(tax_unit, period, parameters)
+        return branch.calculate("eitc_eligible", period)
+
+
 class ca_eitc(Variable):
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/ca/tax/income/credits/earned_income/ca_eitc.py:103]: I think we could now add `defined_for = "ca_eitc_eligible"`. @nikhilwoodruff any issues with `defined_for` + branching?

File: policyengine_us/parameters/gov/states/ca/tax/income/credits/young_child/loss_threshold.yaml [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  2021-01-01: 0
+  2022-01-01: 32_490
+
+metadata:
+  unit: currency-USD
+  period: year
+  name: ca_young_child_loss_threshold
+  label: CA young child tax credit loss threshold
+  reference:
+    - title: Instructions for 2021 CalEITC Form 3514
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/ca/tax/income/credits/young_child/loss_threshold.yaml:12]:
```suggestion
    # The law specifies $30,000, indexed to inflation from 2022 onward.
    # It does not specify the base year, so we use specific values from the forms.
    - title: California Revenue and Taxation Code §17052.1(b)(1)
      href: https://leginfo.legislature.ca.gov/faces/codes_displaySection.xhtml?lawCode=RTC&sectionNum=17052.1
    - title: Instructions for 2021 CalEITC Form 3514
```

File: policyengine_us/parameters/gov/states/ca/tax/income/credits/young_child/loss_threshold.yaml [hunk 6]
```diff
@@ -0,0 +1,15 @@
+description: CA young child tax credit loss threshold.
+values:
+  2021-01-01: 0
+  2022-01-01: 32_490
+
+metadata:
+  unit: currency-USD
+  period: year
+  name: ca_young_child_loss_threshold
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/ca/tax/income/credits/young_child/loss_threshold.yaml:9]:
```suggestion
```

File: policyengine_us/parameters/gov/states/ca/tax/income/credits/young_child/loss_threshold.yaml [hunk 7]
```diff
@@ -0,0 +1,15 @@
+description: CA young child tax credit loss threshold.
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/ca/tax/income/credits/young_child/loss_threshold.yaml:1]:
```suggestion
description: California limits its Young Child Tax Credit to filers with losses not exceeding this value.
```

File: policyengine_us/parameters/gov/states/ca/tax/income/credits/young_child/loss_threshold.yaml [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+values:
+  2021-01-01: 0
+  2022-01-01: 32_490
+
+metadata:
+  unit: currency-USD
+  period: year
+  name: ca_young_child_loss_threshold
+  label: CA young child tax credit loss threshold
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/ca/tax/income/credits/young_child/loss_threshold.yaml:10]: We're inconsistent about labels right now but I'd like to start spelling them out.

They seem to capitalize the credit in public documents.
```suggestion
  label: California Young Child Tax Credit loss threshold
```

File: policyengine_us/variables/gov/states/ca/tax/income/credits/young_child/ca_yctc.py [hunk 9]
```diff
@@ -9,38 +8,52 @@ class ca_yctc(Variable):
...
+        # determine eligibility, which requires both (a) and (b) to be true:
+        # (a) tax unit has at least one CalEITC-qualifying child
+        # (b) tax unit receives CalEITC _OR_ is CalEITC eligible but has losses
+        # ... determine (a)
         meets_age_limit = person("age", period) < p.ineligible_age
         is_qualifying_child_for_caleitc = person(
             "ca_is_qualifying_child_for_caleitc", period
         )
         is_eligible_child = meets_age_limit & is_qualifying_child_for_caleitc
         has_eligible_child = tax_unit.sum(is_eligible_child) > 0
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/ca/tax/income/credits/young_child/ca_yctc.py:32]:
```suggestion
        has_eligible_child = tax_unit.any(is_eligible_child)
```

File: policyengine_us/variables/gov/states/ca/tax/income/credits/young_child/ca_yctc.py [hunk 10]
```diff
@@ -9,38 +8,52 @@ class ca_yctc(Variable):
...
-        increments = np.round_(excess_earnings / p.phase_out.increment, 2)
-        # Round reduction to two decimal places.
-        reduction = min_(
-            p.amount, np.round_(increments * p.phase_out.amount, 2)
+        # ... determine (b) part one
+        gets_caleitc = tax_unit("ca_eitc", period) > 0
+        # ... determine (b) part two
+        is_caleitc_eligible = tax_unit("ca_eitc_eligible", period)
+        # ... ... determine if losses are limited to modest amount
+        federal_gross_income = tax_unit.sum(person("irs_gross_income", period))
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/ca/tax/income/credits/young_child/ca_yctc.py:38]:
```suggestion
        federal_gross_income = add(tax_unit, period, ["irs_gross_income"])
```

File: policyengine_us/variables/gov/states/ca/tax/income/credits/young_child/ca_yctc.py (hunk 9 above)
Comment (martinholmer) [reply policyengine_us/variables/gov/states/ca/tax/income/credits/young_child/ca_yctc.py:32]: Done

File: policyengine_us/variables/gov/states/ca/tax/income/credits/young_child/ca_yctc.py (hunk 10 above)
Comment (martinholmer) [reply policyengine_us/variables/gov/states/ca/tax/income/credits/young_child/ca_yctc.py:38]: Done

File: policyengine_us/parameters/gov/states/ca/tax/income/credits/young_child/loss_threshold.yaml (hunk 7 above)
Comment (martinholmer) [reply policyengine_us/parameters/gov/states/ca/tax/income/credits/young_child/loss_threshold.yaml:1]: Done

File: policyengine_us/parameters/gov/states/ca/tax/income/credits/young_child/loss_threshold.yaml (hunk 6 above)
Comment (martinholmer) [reply policyengine_us/parameters/gov/states/ca/tax/income/credits/young_child/loss_threshold.yaml:9]: Done

File: policyengine_us/parameters/gov/states/ca/tax/income/credits/young_child/loss_threshold.yaml (hunk 8 above)
Comment (martinholmer) [reply policyengine_us/parameters/gov/states/ca/tax/income/credits/young_child/loss_threshold.yaml:10]: Done

File: policyengine_us/parameters/gov/states/ca/tax/income/credits/young_child/loss_threshold.yaml (hunk 5 above)
Comment (martinholmer) [reply policyengine_us/parameters/gov/states/ca/tax/income/credits/young_child/loss_threshold.yaml:12]: Done

File: policyengine_us/variables/gov/states/ca/tax/income/credits/earned_income/ca_eitc.py (hunk 3 above)
Comment (martinholmer) [reply policyengine_us/variables/gov/states/ca/tax/income/credits/earned_income/ca_eitc.py:96]: Done

File: policyengine_us/variables/gov/states/ca/tax/income/credits/earned_income/ca_eitc.py (hunk 2 above)
Comment (martinholmer) [reply policyengine_us/variables/gov/states/ca/tax/income/credits/earned_income/ca_eitc.py:92]: @MaxGhenis said:
> Could you move this to `ca_eitc_eligible.py` and import the branch from here?

What is the syntax for "import the branch from here"?

Comment (martinholmer) [reply policyengine_us/variables/gov/states/ca/tax/income/credits/earned_income/ca_eitc.py:92]: @MaxGhenis,  If you aren't sure how to import a branch into another module, could you merge PR #1897 as it is now.  Then you and @nikhilwoodruff can experiment with putting the `ca_eitc_eligible` variable in its own module.  Let me point out that having two or more variables defined in the same module has been done in PolicyEngineUS from the very beginning.   So, this PR is not out of line in this respect with the existing code base.

Comment (MaxGhenis) [reply policyengine_us/variables/gov/states/ca/tax/income/credits/earned_income/ca_eitc.py:92]: Sounds good. We would like to move all variables into their own folder (#373). I also filed https://github.com/PolicyEngine/policyengine-us/issues/1906 to use `defined_for` here after testing. But merging for now.

**Pull request merged**